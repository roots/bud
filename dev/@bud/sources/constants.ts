export const REGISTRY_NPM = 'https://registry.npmjs.org/'
export const REGISTRY_PROXY = 'http://localhost:4873'
export const PM2_BIN_PATH = './bin/pm2/index.js'
