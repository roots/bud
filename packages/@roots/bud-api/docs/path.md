# bud.path

`bud.path` accepts up to two strings as parameters:

- The first is a location registered to bud. Out-of-the-box it is one of: `project`, `src`, `dist`, `storage`. It is possible for extensions to register additional locations.

- The second is an optional string which will be joined on the path returned from the first param.

Paths can be redefined using [bud.setPath](https://github.com/roots/bud/tree/stable/docs/config/setPath.md).

## Usage

Returns the absolute path to the src directory.

```js
bud.path("src");
```

Returns the absolute path to 'scripts/app.js' (from project root).

```js
bud.path("project", "scripts/app.js");
```
