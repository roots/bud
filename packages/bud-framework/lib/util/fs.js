"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
exports.__esModule = true;
exports.fs = void 0;
var fs_extra_1 = __importDefault(require("fs-extra"));
var path_1 = __importDefault(require("path"));
var fs = {
    path: path_1["default"],
    fs: fs_extra_1["default"],
    existsSync: fs_extra_1["default"].existsSync
};
exports.fs = fs;
//# sourceMappingURL=fs.js.map