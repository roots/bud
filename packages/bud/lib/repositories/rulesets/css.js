"use strict";
exports.__esModule = true;
exports.css = void 0;
var css = function (bud) {
    return bud.hooks.filter('webpack.module.rules.css', {
        test: bud.hooks.filter('webpack.module.rules.css.test', bud.patterns.get('css')),
        exclude: bud.hooks.filter('webpack.module.rules.css.exclude', bud.patterns.get('vendor')),
        use: bud.hooks.filter('webpack.module.rules.css.use', [
            bud.inProduction
                ? bud.uses.get('miniCss')(bud)
                : bud.loaders.get('style'),
            bud.uses.get('css')(bud),
            bud.uses.get('resolveUrl')(bud),
            bud.uses.get('postCss')(bud),
        ])
    });
};
exports.css = css;
//# sourceMappingURL=css.js.map