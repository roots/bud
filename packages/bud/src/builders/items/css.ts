export const ident: Framework.Item['ident'] = 'css'
export const loader: Framework.Item['loader'] = 'css-loader'
export const options: Framework.Item['options'] = {
  importLoaders: 2,
}
