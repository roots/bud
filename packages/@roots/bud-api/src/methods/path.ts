import {Hooks} from '@roots/bud-framework'
import {join} from 'path'

declare module '@roots/bud-framework' {
  interface Framework {
    /**
     * ## path [💁 Fluent]
     *
     * Get a filepath.
     *
     * ### Usage
     *
     * Absolute path to the dist directory:
     *
     * ```js
     * app.path('dist')
     * ```
     *
     * Absolute path to `scripts/app.js` in the dist directory:
     *
     * ```js
     * app.path('dist', 'scripts/app.js')
     *  ```
     */
    path: Api.Path
  }

  namespace Api {
    export {Path}
  }
}

type PathKey = `${keyof Hooks.Locale.Definitions & string}`

type Path = (key: PathKey, path?: string) => string

export const path: Path = function (key, path?) {
  return join(
    ...[
      key !== 'project'
        ? this.hooks.filter('location/project')
        : false,
      this.hooks.filter(`location/${key}`),
      path ?? false,
    ].filter(Boolean),
  )
}
