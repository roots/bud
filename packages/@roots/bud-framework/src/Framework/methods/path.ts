import {join} from 'path'

import {Framework, Locations} from '../..'

interface path {
  (
    this: Framework,
    key: keyof Locations & string,
    ...path: string[]
  ): string
}

interface path {
  (key: keyof Locations & string, ...path: string[]): string
}

const path: path = function (
  key: keyof Locations & string,
  ...path: string[]
): string {
  const project = this.hooks.filter(`location.project`)
  const partial = this.hooks.filter(`location.${key}`)

  const useAbsolute =
    project === partial || partial.startsWith('/')

  return useAbsolute
    ? join(...[partial, ...(path ?? [])].filter(Boolean))
    : join(
        ...[project, partial, ...(path ?? [])].filter(Boolean),
      )
}

export {path}
