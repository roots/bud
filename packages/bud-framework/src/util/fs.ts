import filesystem from 'fs-extra'
import path from 'path'
import {URL} from 'url'

import type {PlatformPath} from 'path'

type PathLike = string | Buffer | URL
type Exists = (path: PathLike) => boolean
type FS = {
  path: PlatformPath
  fs: typeof filesystem
  existsSync: Exists
}

const fs: FS = {
  path,
  fs: filesystem,
  existsSync: filesystem.existsSync,
}

export {fs}
export type {FS}
