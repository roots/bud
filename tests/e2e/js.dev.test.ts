/* eslint-disable no-console */
import {describe, expect, it, jest} from '@jest/globals'
import {paths} from '@repo/constants'
import {execa, ExecaChildProcess} from 'execa'
import fs from 'fs-extra'
import {join} from 'path'
import {Browser, chromium, Page} from 'playwright'

jest.setTimeout(100000)

describe('hmr js', () => {
  let browser: Browser
  let page: Page
  let devProcess: ExecaChildProcess

  beforeAll(done => {
    chromium
      .launch()
      .then(instance => {
        browser = instance
      })
      .then(async () => {
        try {
          devProcess = execa(
            'node',
            ['./node_modules/.bin/bud', 'dev', '--html'],
            {
              cwd: join(paths.mocks, 'yarn', 'basic'),
            },
          )

          devProcess.stdout?.on('data', () => {
            setTimeout(done, 2000).unref()
          })
        } catch (e) {
          console.error(e)
        }
      })
  })

  afterAll(async () => {
    devProcess.kill()
    await browser.close()
  })

  beforeEach(async () => {
    page = await browser.newPage()
    await page.goto('http://0.0.0.0:3000/')
  })

  afterEach(async () => {
    await page.close()
  })

  it('readied up', async () => {
    const title = await page.title()
    expect(title).toBe('Webpack App')
  })

  it('has indicator component', async () => {
    const indicator = await page.$('bud-activity-indicator')

    const warnings = await indicator?.getAttribute('has-warnings')
    expect(warnings).toBe('0')

    const errors = await indicator?.getAttribute('has-errors')
    expect(errors).toBe('0')

    const html = await page?.innerHTML('bud-activity-indicator')
    expect(html).toMatchSnapshot()
  })

  it('hot updates js', async () => {
    await page.waitForTimeout(1000)

    const body = await page.$('body')
    const classList = await body?.getProperty('classList')
    const classes = await classList?.jsonValue()

    expect(Object.values(classes)).toContain('init')
    expect(Object.values(classes)).toMatchSnapshot()

    await fs
      .writeFile(
        join(paths.mocks, 'yarn', 'basic', 'src', 'index.js'),
        `\
import './styles.css'

document.querySelector('body')?.classList.add('hot')

module?.hot?.accept()
`,
      )
      .then(async () => {
        await page.waitForTimeout(1000)

        const body = await page.$('body')
        const classList = await body?.getProperty('classList')
        const classes = await classList?.jsonValue()
        expect(Object.values(classes)).toContain('hot')
        expect(Object.values(classes)).toMatchSnapshot()
      })
  })
})
