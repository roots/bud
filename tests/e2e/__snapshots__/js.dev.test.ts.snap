// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hmr js has indicator component 1`] = `""`;

exports[`hmr js hot updates js 1`] = `
Array [
  "init",
]
`;

exports[`hmr js hot updates js 2`] = `
Array [
  "init",
  "hot",
]
`;
